client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -118,22 +109,21 @@
         Exceptions.checkNotClosed(closed.get(), this);
         ByteBuffer data = serializer.serialize(event);
         CompletableFuture<Void> ackFuture = new CompletableFuture<Void>();
-        writeFlushLock.writeLock().lock();
-        try {
-            SegmentOutputStream segmentWriter = selector.getSegmentOutputStreamForKey(routingKey);
-            while (segmentWriter == null) {
-                log.info("Don't have a writer for segment: {}", selector.getSegmentForEvent(routingKey));
-                handleMissingLog();
-                segmentWriter = selector.getSegmentOutputStreamForKey(routingKey);
+        synchronized (writeFlushLock) {
```

Comment:
Why do we synchronize in lines 112 and 113, and not right before line 120? That would reduce the scope of the locks, but I'm not sure there is a concurrency concern with respect to getting the segment writer.

Reply:
All those classes are threadsafe, so yes, there is no concurrency concern. The reason for the locks is to ensure that:
a) The segments don't change while the write is occurring. (This could result in a 'late write' to a sealed segment, resulting in an effectively lost event)
b) ensuring that two write calls have a defined order.
The reason writeSealLock needs to be held starting on 113 is to insure (a). The reason the writeFlushLock is held on 112 and not just inside on 120 is because it needs to be the outer lock with the flush lock being the inner one.